source/server/config_validation/connection.h

```diff
@@ -0,0 +1,27 @@
+#include "common/network/connection_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+/*
+  Class representing connection to upstream entity in config verification mode. The connection is
+  not really established, but sockets may be allocated. Methods doing "real" connections should be
+  overridden with no-op implementations.
+*/
+class ConfigValidateConnection : public Network::ClientConnectionImpl {
```

Comment:
Correct me if I'm wrong, but it seems like it would be easy to start doing real connections on the socket without intending it? (This concerns me more than theoretically as I think the intention of validation mode is to have no observable effects). I realize it would be tedious, but should we instead have a validation version of connection which does not derive from ClientConnectionImpl and potentially NOPs whatever needs to be "mocked?"

Reply:
@mattklein123 You are correct. It is possible that certain configuration item may take a process a step further and call "connect" method. This is exactly what happened in #3083. Presence of health check triggers a call to Dispatcher::createClientConnection.
To take away risk of opening unintended connections upstream I see few options:
- as you suggest: derive ConfigValidateConnection from pure ClientConnection and mock what is required leaving rest of methods no-op. ClientConnection has rather large number of methods though. And the risk is that certain configuration item may move the process a step further and require that ClientConnection methods return a valid thing. For lack of better example, it may ask for remote address. Also, this approach will start parallel hierarchy of classes.
- patch unwanted methods. This is what was originally done with the dispatcher: https://github.com/envoyproxy/envoy/blob/master/source/server/config_validation/dispatcher.h#L15 and https://github.com/envoyproxy/envoy/blob/master/source/server/config_validation/dispatcher.cc#L11. In this case I would patch ConfigValidationConnection::connect (https://github.com/envoyproxy/envoy/blob/master/include/envoy/network/connection.h#L249) as NOT_IMPLEMENTED to indicate that it needs better handle.
- redefine socket layer to do nothing, but to behave as real socket.  Probably the hardest approach, but cleanest. Dispatcher, clientConnection, etc objects would require no changes.

Reply:
I don't have a strong opinion. I guess either option 2 or 3? WDYT?

Reply:
I would suggest to go with option 2 for now: patch it with NOT_IMPLEMENTED. It fixes the reported bug. After that I can spend few days experimenting with fake socket layer and share my findings. Also, I would suggest to have a directory with many configs files and run them in config verification mode as parameterized unit tests.

Reply:
Sure SGTM. re: config tests, we already have a ton (see config_test). It would be good to additionally run all of them through verification mode. Can you take a look at that?

Reply:
I will finish this PR with option 2 - NOT IMPLEMENTED. Then will experiment with option 3. re: config tests - no problem - I will add procedure to run config verification for all config files in that directory.

Reply:
I had to patch "connect" method to have no observable effects. By tracing what happens on the wire I saw TCP handshakes being started, but they never completed probably because main thread exited and closed all opened sockets. The current code fixes original issue #3083.
I will experiment with option 3 (described above) but it will be separate PR.